app/models/manageiq/providers/azure/cloud_manager.rb

```diff
@@ -114,4 +114,56 @@ def vm_reboot_guest(vm, _options = {})
   rescue => err
     _log.error "vm=[#{vm.name}], error: #{err}"
   end
+
+  def vm_create_evm_snapshot(vm, options = {})
+    conf     = connect(options)
+    vm_svc   = vm.provider_service(conf)
+    snap_svc = snapshot_service(conf)
+    vm_obj   = vm_svc.get(vm.name, vm.resource_group)
+    return unless vm_obj.managed_disk?
+    os_disk      = vm_obj.properties.storage_profile.os_disk
+    snap_options = { :location   => vm.location,
+                     :properties => {
+                       :creationData => {
+                         :createOption     => "Copy",
+                         :sourceResourceId => os_disk.managed_disk.id
+                       }
+                     } }
+    snap_name = os_disk.name + "__EVM__SSA__SNAPSHOT"
+    _log.debug("vm=[#{vm.name}] creating SSA snapshot #{snap_name}")
+    begin
+      snap_svc.get(snap_name, vm.resource_group)
+    rescue ::Azure::Armrest::NotFoundException, ::Azure::Armrest::ResourceNotFoundException => err
+      begin
+        snap_svc.create(snap_name, vm.resource_group, snap_options)
+        return snap_name
+      rescue => err
+        _log.error("vm=[#{vm.name}], error: #{err}")
+        _log.debug { err.backtrace.join("\n") }
+        raise "Error #{err} creating SSA Snapshot #{snap_name}"
+      end
+    end
+    _log.error("SSA Snapshot #{snap_name} already exists.")
+    raise "Snapshot #{snap_name} already exists. Another SSA request for this VM is in progress or a previous one failed to clean up properly."
```

Comment:
Who calls this method? I'm seeing the possibility of a caller having to react to two different error conditions in possibly two different ways. One error condition is when it can't create the snapshot because of an error (probably in Azure). And another error condition when an SSA request is already being processed for this VM.

Maybe the "snapshot already exists" situation could be a specific error type? Then, the caller could rescue that error type in its own block with a special handler. And all other errors could be assumed to be an actual failure to process the request by the provider.

I could also see the "snapshot already exists" situation being a non-error, and just log a warning and return nil. But, that might be too ambiguous for the caller.

Thoughts?

Reply:
@blomquisg It is called by VmScan.create_snapshot. We have similar logic for the Scvmm code as well (although some of it is embedded in the manageiq-smartstate repo. 

It really doesn't matter to VmScan if we can't perform the snapshot because of a provider error or because the snapshot already exists - the SSA request needs to fail in either case.

The "snapshot already exists" situation cannot be a non-error - if there really is another request to run SSA on the same instance/image that we somehow missed (which should not happen) a problem would arise when the first request finished and therefore deleted the snapshot.  Since these requests are stateless and the snapshot disk isn't actually mounted or opened the second request would then run into an unexplained error.

Let me know what you think.  @roliveri, feel free to chime in if you feel otherwise.

Reply:
Historically, `vm_create_evm_snapshot` is not a general purpose snapshot creation method, `evm` snapshots are only created for SSA. SSA doesn't allow multiple `evm` (SSA) snapshots, hence the error.

Maybe there should be a general `create_snapshot` method, that is called by `vm_create_evm_snapshot`.

Reply:
@roliveri you mean as a follow-on task for *all* the providers needing it, right?

Reply:
Yes. If a provider needs a general create snapshot facility, then it makes sense to structure the code that way. As it stands now, `vm_create_evm_snapshot` is SSA specific, so it should be ok the way it is.